Integrate position-area and self-alignment properties (align-self/justify-self)

fantasai opened the pull request:
#### b2c8205237ab80de04f20ae98687373a645ca27f
<pre>
Integrate position-area and self-alignment properties (align-self/justify-self)
<a href="https://bugs.webkit.org/show_bug.cgi?id=288886">https://bugs.webkit.org/show_bug.cgi?id=288886</a>
<a href="https://rdar.apple.com/145889235">rdar://145889235</a>

Reviewed by Antti Koivisto.

Incorporates position-area defaults into alignment resolution by adding
a method to return the default alignment in PositionArea and incorporating
that into positioned layout code through new alignment resolution methods
on PositionedLayoutConstraints.

Also caches RenderStyle into PositionedLayoutConstraints, because we need
it during alignment.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-014-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/auto-margins-position-area-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-fallbacks-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-try-rule-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-intermediate-ignored-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-align-justify-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-align-justify-wm-dir-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-in-position-try-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-scrolling-003.tentative-expected.txt:
* Source/WebCore/rendering/RenderBox.cpp: Cache RenderStyle into PositionedLayoutConstraints.
(WebCore::RenderBox::PositionedLayoutConstraints::needsAnchor const):
(WebCore::RenderBox::PositionedLayoutConstraints::PositionedLayoutConstraints):
(WebCore::RenderBox::PositionedLayoutConstraints::captureInsets):
(WebCore::RenderBox::PositionedLayoutConstraints::computeAnchorGeometry):
(WebCore::RenderBox::PositionedLayoutConstraints::adjustForPositionArea):
* Source/WebCore/rendering/RenderBox.cpp: Integrate position-area default alignment.
(WebCore::RenderBox::PositionedLayoutConstraints::resolveAlignmentAdjustment const): Update to use new alignment resolution method.
(WebCore::RenderBox::PositionedLayoutConstraints::resolveAlignmentPosition const): Add alignment resolution method.
(WebCore::RenderBox::PositionedLayoutConstraints::alignmentAppliesStretch const): Add alignment resolution method.
(WebCore::RenderBox::computePositionedLogicalWidth const): Update alignment check to use resolution method.
(WebCore::RenderBox::computePositionedLogicalWidthUsing const): Update alignment check to use resolution method.
(WebCore::RenderBox::computePositionedLogicalHeight const): Update alignment check to use resolution method.
(WebCore::RenderBox::computePositionedLogicalHeightUsing const): Update alignment check to use resolution method.
* Source/WebCore/rendering/style/PositionArea.cpp:
(WebCore::PositionArea::spanForAxis const): Add helper method.
(WebCore::PositionArea::coordMatchedTrackForAxis const): Update to use helper method.
(WebCore::flip): Add helper function.
(WebCore::PositionArea::defaultAlignmentForAxis const): Implement method to return default alignment.
* Source/WebCore/rendering/style/PositionArea.h:

Canonical link: <a href="https://commits.webkit.org/291552@main">https://commits.webkit.org/291552@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c0fcc5706a5367986d2a5a75794b26eb953f088b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93195 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12750 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2557 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98194 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43721 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95245 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13029 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21203 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71242 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28641 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-anchor-position/position-visibility-remove-anchors-visible.html (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96197 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9793 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84365 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51574 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9488 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/1999 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43035 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79777 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2009 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100223 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20247 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14829 "Found 1 new test failure: ipc/large-vector-allocate-failure-crash.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80257 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20499 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80285 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79566 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19782 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24116 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/1488 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13362 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20231 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25408 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19918 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23378 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21659 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on LayoutTests/TestExpectations:
```diff
@@ -7598,26 +7598,21 @@ imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-update-007
 imported/w3c/web-platform-tests/css/css-anchor-position/position-visibility-anchors-visible-chained-004.html [ ImageOnlyFailure ]
-imported/w3c/web-platform-tests/css/css-anchor-position/position-visibility-anchors-visible-change-anchor.html [ ImageOnlyFailure ]
 imported/w3c/web-platform-tests/css/css-anchor-position/position-visibility-anchors-visible-change-css-visibility.html [ ImageOnlyFailure ]
-imported/w3c/web-platform-tests/css/css-anchor-position/position-visibility-anchors-visible-css-visibility.html [ ImageOnlyFailure ]
```
looks like this one is not passing in the bots yet

fantasai committed:
Integrate position-area and self-alignment properties (align-self/justify-self)
https://bugs.webkit.org/show_bug.cgi?id=288886
rdar://145889235

Reviewed by Antti Koivisto.

Incorporates position-area defaults into alignment resolution by adding
a method to return the default alignment in PositionArea and incorporating
that into positioned layout code through new alignment resolution methods
on PositionedLayoutConstraints.

Also caches RenderStyle into PositionedLayoutConstraints, because we need
it during alignment.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-014-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/auto-margins-position-area-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-fallbacks-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-try-rule-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-intermediate-ignored-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-align-justify-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-align-justify-wm-dir-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-in-position-try-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-scrolling-003.tentative-expected.txt:
* Source/WebCore/rendering/RenderBox.cpp: Cache RenderStyle into PositionedLayoutConstraints.
(WebCore::RenderBox::PositionedLayoutConstraints::needsAnchor const):
(WebCore::RenderBox::PositionedLayoutConstraints::PositionedLayoutConstraints):
(WebCore::RenderBox::PositionedLayoutConstraints::captureInsets):
(WebCore::RenderBox::PositionedLayoutConstraints::computeAnchorGeometry):
(WebCore::RenderBox::PositionedLayoutConstraints::adjustForPositionArea):
* Source/WebCore/rendering/RenderBox.cpp: Integrate position-area default alignment.
(WebCore::RenderBox::PositionedLayoutConstraints::resolveAlignmentAdjustment const): Update to use new alignment resolution method.
(WebCore::RenderBox::PositionedLayoutConstraints::resolveAlignmentPosition const): Add alignment resolution method.
(WebCore::RenderBox::PositionedLayoutConstraints::alignmentAppliesStretch const): Add alignment resolution method.
(WebCore::RenderBox::computePositionedLogicalWidth const): Update alignment check to use resolution method.
(WebCore::RenderBox::computePositionedLogicalWidthUsing const): Update alignment check to use resolution method.
(WebCore::RenderBox::computePositionedLogicalHeight const): Update alignment check to use resolution method.
(WebCore::RenderBox::computePositionedLogicalHeightUsing const): Update alignment check to use resolution method.
* Source/WebCore/rendering/style/PositionArea.cpp:
(WebCore::PositionArea::spanForAxis const): Add helper method.
(WebCore::PositionArea::coordMatchedTrackForAxis const): Update to use helper method.
(WebCore::flip): Add helper function.
(WebCore::PositionArea::defaultAlignmentForAxis const): Implement method to return default alignment.
* Source/WebCore/rendering/style/PositionArea.h:

Canonical link: https://commits.webkit.org/291552@main
